### PR TITLE
Codechange: Use Map::Iterate() to iterate tiles

### PR DESCRIPTION
**Pull request by SamuXarick**
## Motivation / Problem

- Some whole map tile iterations are still using `for (TileIndex t = 0; t < Map::Size(); t++)`
- Not using `const` on those that already use `Map::Iterate()`

## Description

- Replace most `TileIndex` map iterations with `for (const auto t : Map::Iterate())`
- Use `const` when possible

## Limitations
There are still a few iterations that still use `tile++` for iterating, and many specialized stuff in saveload code which I better not touch.

## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR touches english.txt or translations? Check the [guidelines](https://github.com/OpenTTD/OpenTTD/blob/master/docs/eints.md)
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, game_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

### CODE REVIEW

**Comment by SamuXarick** [comment]
Replaced a `while` loop with `Map::Iterate()`

**Comment by SamuXarick** [comment]
Created a 3rd commit with another conversion during waypoint conversion.

**Comment by LordAro** [comment]
It's very subjective, but I tend to prefer being explicit with types for integer types, which TileIndex is (well, sort of, it's a strongtype'd integer). It's a short type name anyway.

Some others may disagree

File: src/landscape.cpp [hunk 1]
```diff
@@ -948,8 +948,8 @@ static void CreateDesertOrRainForest(uint desert_tropic_line)
 {
 	uint update_freq = Map::Size() / 4;
 
-	for (TileIndex tile = 0; tile != Map::Size(); ++tile) {
-		if ((tile.base() % update_freq) == 0) IncreaseGeneratingWorldProgress(GWP_LANDSCAPE);
```
Comment (2TallTyler) [inline src/landscape.cpp:954]: I don't think you should change this `.base()` here. Make a separate PR if you must.

**Review by 2TallTyler — COMMENTED** [review]
~~I also prefer explicit types when they're simple and short. `TileIndex` is indeed not much longer than `auto`, and easier for me to read. (Yes, I know IDEs can show the auto type when I hover....but I don't want to hover over everything!)~~

~~We currently have both `TileIndex` and `auto`, so I think it's just a question of which standard to follow going forward. I do not claim to know best.~~

**Comment by PeterN** [comment]
> I also prefer explicit types when they're simple and short. `TileIndex` is indeed not much longer than `auto`, and easier for me to read. (Yes, I know IDEs can show the auto type when I hover....but I don't want to hover over everything!)
> 
> We currently have both `TileIndex` and `auto`, so I think it's just a question of which standard to follow going forward. I do not claim to know best.

`Map::Iterate()` has ALWAYS been used with `auto` so far. Also it deals with `Tile`, not `TileIndex`.

File: src/landscape.cpp (hunk 1 above)
Comment (SamuXarick) [reply src/landscape.cpp:954]: I can make it into its own commit. Doesn't need to be a new PR imo.

**Review by 2TallTyler — APPROVED** [review]
